lxd/instance/drivers/driver_qemu.go

```diff
@@ -4200,6 +4209,16 @@ func (d *qemu) addNetDevConfig(busName string, qemuDev map[string]string, bootIn
 				}
 			}
 
+			defer func() {
```

Comment:
what linter suggestion is this trying to resolve?

Reply:
What ever the suggestion, the fix here is invalid, because by delaying the definition of the defer you're introducing the possibility for LXD to leak file handles if the function returns early on an error.

Reply:
The linterr suggested to take the defer blocks out of the loop. If this is the only way, is there a comment syntax to avoid the lint on this line then?

Reply:
why did it suggest that?

Reply:
I found this which could be an explanation on why a defer inside a loop causes this warning:
https://stackoverflow.com/questions/45617758/proper-way-to-release-resources-with-defer-in-a-loop#:~:text=The%20whole%20point,without%20defer%3A

Reply:
I think the original implementation was correct, as the comment says:

>  // Close file after device has been added.

So if we close it inside the for loop thats too earlier as `err = m.AddNIC(qemuNetDev, qemuDev)` has not been called yet.

But we do want to close them after `err = m.AddNIC(qemuNetDev, qemuDev)` has been called, which is just before it returns and so the defers will be called, or in case there is an error we also want them to be closed.

Reply:
Would this approach be ok in order to avoid the leak ?:
```go
{
        ...
        openedFiles := make([]io.Closer, 0, queueCount*2)
        closeFiles := func(files []io.Closer) {
	        for _, f := range files {
		        _ = f.Close()
	        }
        }

	for i := 0; i < queueCount; i++ {
		devFile, err := deviceFile()
		if err != nil {
			closeFiles(openedFiles)
			return fmt.Errorf("Error opening netdev file for queue %d: %w", i, err)
		}

		openedFiles = append(openedFiles, devFile)
		devFDName := fmt.Sprintf("%s.%d", devFile.Name(), i)
		err = m.SendFile(devFDName, devFile)
		if err != nil {
			closeFiles(openedFiles)
			return fmt.Errorf("Failed to send %q file descriptor for queue %d: %w", devFDName, i, err)
		}

		reverter.Add(func() { _ = m.CloseFile(devFDName) })

		fds = append(fds, devFDName)

		if vhostNetEnabled {
			// Open a vhost-net file handle for each device file handle.
			vhostFile, err := os.OpenFile("/dev/vhost-net", os.O_RDWR, 0)
			if err != nil {
				closeFiles(openedFiles)
				return fmt.Errorf("Error opening /dev/vhost-net for queue %d: %w", i, err)
			}

			openedFiles = append(openedFiles, vhostFile)
			vhostFDName := fmt.Sprintf("%s.%d", vhostFile.Name(), i)
			err = m.SendFile(vhostFDName, vhostFile)
			if err != nil {
				closeFiles(openedFiles)
				return fmt.Errorf("Failed to send %q file descriptor for queue %d: %w", vhostFDName, i, err)
			}

			reverter.Add(func() { _ = m.CloseFile(vhostFDName) })

			vhostfds = append(vhostfds, vhostFDName)
		}
	}

	qemuNetDev := map[string]any{
		"id":    fmt.Sprintf("%s%s", qemuNetDevIDPrefix, escapedDeviceName),
		"type":  "tap",
		"vhost": vhostNetEnabled,
	}

	if shared.ValueInSlice(busName, []string{"pcie", "pci"}) {
		qemuDev["driver"] = "virtio-net-pci"
	} else if busName == "ccw" {
		qemuDev["driver"] = "virtio-net-ccw"
	}

	qemuNetDev["fds"] = strings.Join(fds, ":")

	if len(vhostfds) > 0 {
		qemuNetDev["vhostfds"] = strings.Join(vhostfds, ":")
	}

	qemuNetDevID, ok := qemuNetDev["id"].(string)
	if !ok {
		return fmt.Errorf("Failed getting QEMU netdev id")
	}

	qemuDev["netdev"] = qemuNetDevID
	qemuDev["mac"] = devHwaddr

	err = m.AddNIC(qemuNetDev, qemuDev)
	if err != nil {
		return fmt.Errorf("Failed setting up device %q: %w", devName, err)
	}

	// Cleanup resources
	closeFiles(openedFiles)
	...
}
```

Reply:
Lets just use defer with a lint ignore rule here. Rather than adding additional cognitive overhead to keep track of it.

The rule itself says "may not run" but in this case it does run when we expect them to.

Reply:
ok